### PR TITLE
Safeguarding against usage of uninitialized DistributedManager

### LINKED ISSUE #474: 🐛[BUG]: DistributedManager gets silently initialized as a single process job if instantiated before initializing
### Version

main

### On which installation method(s) does this occur?

Source

### Describe the issue

This works as expected:
```python
In [1]: from modulus.distributed import DistributedManager

In [2]: DistributedManager.is_initialized()
Out[2]: False

In [3]: DistributedManager.initialize()

In [4]: DistributedManager.is_initialized()
Out[4]: True

In [5]: manager = DistributedManager()

In [6]: manager._initialization_method
Out[8]: 'None'
```
but this does not:
```python
  In [1]: from modulus.distributed import DistributedManager                                                                                                                                                                                                                                                                    
                                                                                                                                                                                                                                                                                                                              
  In [2]: manager = DistributedManager()                                                                                                                                                                                                                                                                                        
                                                                                                                                                                                                                                                                                                                                
  In [3]: manager._initialization_method                                                                                                                                                                                                                                                                                        
  Out[3]: 'None'                                                                                                                                                                                                                                                                                                                
                                                                                                                                                                                                                                                                                                                                
  In [4]: manager.is_initialized()                                                                                                                                                                                                                                                                                              
  Out[4]: True     
```

### Minimum reproducible example

```shell
In [1]: from modulus.distributed import DistributedManager                                                                                                                                                                                                                                                                    
                                                                                                                                                                                                                                                                                                                               
  In [2]: manager = DistributedManager()                                                                                                                                                                                                                                                                                        
                                                                                                                                                                                                                                                                                                                                
  In [3]: manager._initialization_method                                                                                                                                                                                                                                                                                        
  Out[3]: 'None'                                                                                                                                                                                                                                                                                                                
                                                                                                                                                                                                                                                                                                                                
  In [4]: manager.is_initialized()                                                                                                                                                                                                                                                                                              
  Out[4]: True                                                                                                                                                                                                                                                                                                                  
                                                                                                                                                                                                                                                                                                                                
  In [5]: manager.initialize()                                                                                                                                                                                                                                                                                                  
  /code/modulus-core/modulus/distributed/manager.py:302: UserWarning: Distributed manager is already intialized                                                                                                                                                                                                                 
    warn("Distributed manager is already intialized")
```


### Relevant log output

_No response_

### Environment details

_No response_

### PR DESCRIPTION
**Pull request by akshaysubr**
# Modulus Pull Request

## Description

closes #474 

Should be merged in after #469 

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/modulus/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/modulus/blob/main/CHANGELOG.md) is up to date with these changes.
- [x] An [issue](https://github.com/NVIDIA/modulus/issues) is linked to this pull request.

## Dependencies

None

### CODE REVIEW

File: test/distributed/test_manager.py [hunk 1]
```diff
@@ -177,6 +182,46 @@ def test_manager_singleton():
...
+    del os.environ["WORLD_SIZE"]
+
+
+def test_manager_undefined_group_query():
+    os.environ["MASTER_ADDR"] = "localhost"
+    os.environ["MASTER_PORT"] = "12345"
+    os.environ["RANK"] = "0"
+    os.environ["WORLD_SIZE"] = "1"
+    DistributedManager.initialize()
+    print(DistributedManager())
```
Comment (Alexey-Kamenev) [inline test/distributed/test_manager.py:207]: Do you really need `print` here?

File: modulus/distributed/manager.py [hunk 2]
```diff
@@ -26,6 +26,32 @@
...
+class ModulusUninitializedDistributedManagerWarning(Warning):
+    """Warning to indicate usage of an uninitialized DistributedManager"""
+
+    def __init__(self):
+        message = "Instantiating DistributedManager before calling DistributedManager.initialize is not recommended"
+        super().__init__(message)
+
+
 class DistributedManager(object):
     """Distributed Manager for setting up distributed training enviroment.
```
Comment (Alexey-Kamenev) [inline modulus/distributed/manager.py:56]: "enviroment" -> "environment" here and in other places.

**Review by Alexey-Kamenev — APPROVED** [review]
LGTM

**Review by NickGeneva — APPROVED** [review]

File: modulus/distributed/manager.py (hunk 2 above)
Comment (akshaysubr) [reply modulus/distributed/manager.py]: Fixed

File: test/distributed/test_manager.py (hunk 1 above)
Comment (akshaysubr) [reply test/distributed/test_manager.py]: Good point. Removed.

File: modulus/distributed/manager.py [hunk 3]
```diff
@@ -44,8 +44,16 @@ def __init__(self, name: str):
         super().__init__(message)
 
 
+class ModulusUninitializedDistributedManagerWarning(Warning):
+    """Warning to indicate usage of an uninitialized DistributedManager"""
+
+    def __init__(self):
+        message = "Instantiating DistributedManager before calling DistributedManager.initialize is not recommended"
```
[the remark this answers is not in the document]
Comment (akshaysubr) [reply modulus/distributed/manager.py]: Having a more verbose message is a good idea. Made this change.

**Review by dallasfoster — APPROVED** [review]

**Comment by azrael417** [comment]
What it does makes sense. Is this fix supposed to safeguard against initialized DM but requesting an uninitialized distributed group?

**Comment by akshaysubr** [comment]
@azrael417 Not quite. This PR is safeguarding against using the manager before calling `DistributedManager.initialize()` first. There was a bug in CorrDiff where this was silently happening causing a multi GPU job to behave like independent single GPU jobs since that's the default.

**Pull request merged**